### PR TITLE
Asyncapi

### PR DESCRIPTION
**Pull request by keithralphs**
Initial version for review, please see if it looks sensible

### CODE REVIEW

**Review by callumforrester — DISMISSED** [review]
Branches were a bit mangled so probably needs a rebase

File: asyncapi.yaml [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    name: Callum Forrester
+    email: callum.forrester@diamond.ac.uk
+servers:
+  p45-development:
+    url: tcp://p45-activemq.diamond.ac.uk
+    description: Development Test Beamline (Located in Diamond House G13)
+    protocol: JMS
+defaultContentType: application/json
+channels:
+  public/worker/event:
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:17]: According to the spec you're supposed to use the [RFC 6570 URI template](https://tools.ietf.org/html/rfc6570), which implies slashes, but dots do work so I'm not sure which is best.

File: asyncapi.yaml [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    workerStateEvent:
+      type: object
+      required:
+        - state
+        - errors
+        - warnings
+      properties:
+        state:
+          $ref: "#/components/schemas/workerState"
+        task_status:
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:189]: Done

**Review by DiamondJoseph — DISMISSED** [review]

File: asyncapi.yaml [hunk 3]
```diff
@@ -0,0 +1,295 @@
+asyncapi: "2.6.0"
+id: https://github.com/DiamondLightSource/blueapi
+info:
+  title: Bluesky Worker Service Asynchronous API
+  version: 0.2.2
```
[the remark this answers is not in the document]
Comment (callumforrester) [reply asyncapi.yaml:5]: Or a job that automatically updates this based on the git tag, which is how the rest of the tooling works.

File: asyncapi.yaml [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+  description: Service for controlling access to and running scans based on Bluesky Plans and Ophyd Devices
+  contact:
+    name: Callum Forrester
+    email: callum.forrester@diamond.ac.uk
+servers:
+  p45-development:
+    url: tcp://p45-activemq.diamond.ac.uk
+    description: Development Test Beamline (Located in Diamond House G13)
+    protocol: JMS
```
[the remark this answers is not in the document]
Comment (callumforrester) [reply asyncapi.yaml:14]: Ooh, that sounds nice

File: asyncapi.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (callumforrester) [reply asyncapi.yaml:17]: @DiamondJoseph are there any restrictions on this in the rabbit case?

File: asyncapi.yaml [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+      required:
+        - destination
+      properties:
+        destination:
+          type: string
+        reply_destination:
+          type: string
+        correlation_id:
+          type: string
+    runPlan:
```
[the remark this answers is not in the document]
Comment (callumforrester) [reply asyncapi.yaml:145]: Arbitrary params to metadata is currently essential

File: asyncapi.yaml [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+      message:
+        messageId: runPlanResponse
+        headers:
+          $ref: "#/components/schemas/contextHeaders"
+        payload:
+          $ref: "#/components/schemas/taskResponse"
+  worker/plans:
+    description: >-
+      Accepts requests to return the names of all the Bluesky Plans known to the service. At the moment 
+      ny message on this channe will produce this response
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:87]: done

File: asyncapi.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (callumforrester) [reply asyncapi.yaml:189]: There are a few others in the file

File: asyncapi.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (DiamondJoseph) [reply asyncapi.yaml:17]: foo.bar.baz seems to be the preferred

File: src/blueapi/service/app.py [hunk 7]
```diff
@@ -14,6 +17,13 @@
 logging.basicConfig(level=logging.INFO)
 
 
+spec = AutoSpec(
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply src/blueapi/service/app.py:20]: Doesn't seem to be there in my branch

File: asyncapi.yaml [hunk 8]
```diff
@@ -0,0 +1,295 @@
+asyncapi: "2.6.0"
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:1]: moved to docs/user/reference and referenced from both user/reference/index.rst and developer/reference/index.rst

File: asyncapi.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:17]: will use dots. Done

File: asyncapi.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:189]: also done

File: asyncapi.yaml [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+          $ref: "#/components/schemas/workerProgressEvent"
+  public/worker/event/data:
+    description: Produces events supplying data produced by runs associated with a plan
+    subscribe:
+      operationId: dataEvent
+      summary: Events representing collection of data
+      description: Data documents produced by running plans
+      message:
+        messageId: blueskyDoc
+        payload:
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:49]: Done now including local start doc definition until the one on the web is fixed

File: asyncapi.yaml [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+      summary: Submit a 'Run Plan' task with parameters
+      description: Consume a request to run a scan based on the specified Plan with the supplied parameters
+      message:
+        messageId: RunPlan
+        headers:
+          $ref: "#/components/schemas/contextHeaders"
+        payload:
+          $ref: "#/components/schemas/runPlan"
+    subscribe:
+      operationId: runPlanResponse
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:75]: We don't know how to do this at the moment in asyncapi and these exchanges are to be rep;laced by a REST API so for the moment the response definition remains under the calling channel along with a note explaining the use of the temporary one named as per replyDestination in the originating header.

File: asyncapi.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:5]: would be nice but probably for a later incarnation

File: asyncapi.yaml [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        - name
+        - params
+      properties:
+        name:
+          description: Name of plan to run
+          type: string
+        params:
+          description: Values for parameters to plan, if any (Map of name to any)
+          type: object
+          additionalProperties: true
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:157]: This was just a shorthand for a map of string to object i.e.
          additionalProperties:
            type: object
so I've replaced it with that as it conveys more

File: asyncapi.yaml [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+    taskResponse:
```
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:158]: done

File: asyncapi.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:145]: I'll leave as is for now then

File: asyncapi.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (keithralphs) [reply asyncapi.yaml:14]: would be nice but probably for a later incarnation. For the moment will just use a placeholder

[the remark this answers is not in the document]
Comment (callumforrester) [reply asyncapi.yaml:14]: Documented in https://github.com/DiamondLightSource/blueapi/issues/155

File: asyncapi.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (callumforrester) [reply asyncapi.yaml]: Documented in https://github.com/DiamondLightSource/blueapi/issues/156

**Review by callumforrester — APPROVED** [review]
LGTM

**Review by callumforrester — APPROVED** [review]
SLGTM

**Pull request merged**